pkg/manifest/objects.go

```diff
@@ -0,0 +1,326 @@
+	}
+
+	u, ok := o.(*unstructured.Unstructured)
+	if !ok {
+		return nil, fmt.Errorf("parsed unexpected type %T", o)
+	}
+
+	return &Object{
+		object: u,
+		Group:  gvk.Group,
+		Kind:   gvk.Kind,
+		Name:   u.GetName(),
+		json:   json,
+	}, nil
+}
+
+// ParseYAMLToObject parses YAML to an Object.
+func ParseYAMLToObject(yaml []byte) (*Object, error) {
+	r := bytes.NewReader([]byte(yaml))
+	decoder := k8syaml.NewYAMLOrJSONDecoder(r, 1024)
+
+	out := &unstructured.Unstructured{}
+	err := decoder.Decode(out)
+	if err != nil {
+		return nil, fmt.Errorf("error decoding object: %v", err)
+	}
+	return NewObject(out, nil, []byte(yaml)), nil
+}
+
+// UnstructuredObject exposes the raw object, primarily for testing
+func (o *Object) UnstructuredObject() *unstructured.Unstructured {
+	return o.object
+}
+
+// GroupKind returns the GroupKind for o.
+func (o *Object) GroupKind() schema.GroupKind {
+	return o.object.GroupVersionKind().GroupKind()
+}
+
+// GroupVersionKind returns the GroupVersionKind for o.
+func (o *Object) GroupVersionKind() schema.GroupVersionKind {
+	return o.object.GroupVersionKind()
+}
+
+// Hash returns a unique hash for o.
+func (o *Object) Hash() string {
+	return Hash(o.Kind, o.Namespace, o.Name)
+}
+
+// Hash returns a hash for o based on name and kind only.
+func (o *Object) HashNameKind() string {
+	return HashNameKind(o.Kind, o.Name)
+}
+
+// JSON returns a JSON representation of o, using an internal cache.
+func (o *Object) JSON() ([]byte, error) {
+	if o.json != nil {
+		return o.json, nil
+	}
+
+	b, err := o.object.MarshalJSON()
+	if err != nil {
+		return nil, err
+	}
+	o.json = b
+	return b, nil
+}
+
+// YAML returns a YAML representation of o, using an internal cache.
+func (o *Object) YAML() ([]byte, error) {
+	if o.yaml != nil {
+		return o.yaml, nil
+	}
+	// TODO: there has to be a better way.
+	oj, err := o.JSON()
+	if err != nil {
+		return nil, err
+	}
+	o.json = oj
+	y, err := yaml.JSONToYAML(oj)
+	if err != nil {
+		return nil, err
+	}
+	o.yaml = y
+	return y, nil
+}
+
+// YAML returns a YAML representation of o, or an error string if the object cannot be rendered to YAML.
+func (o *Object) YAMLDebugString() string {
+	y, err := o.YAML()
+	if err != nil {
+		return fmt.Sprint(err)
+	}
+	return string(y)
+}
+
+// AddLabels adds labels to o.
+func (o *Object) AddLabels(labels map[string]string) {
+	merged := make(map[string]string)
+	for k, v := range o.object.GetLabels() {
+		merged[k] = v
+	}
+
+	for k, v := range labels {
+		merged[k] = v
+	}
+
+	o.object.SetLabels(merged)
+	// Invalidate cached json
+	o.json = nil
+	o.yaml = nil
+}
+
+// Objects holds a collection of objects, so that we can filter / sequence them
+type Objects struct {
+	Items []*Object
+}
+
+// ParseObjectsFromYAMLManifest returns an Objects represetation of manifest.
+func ParseObjectsFromYAMLManifest(manifest string) (*Objects, error) {
+	var b bytes.Buffer
+
+	var yamls []string
+	for _, line := range strings.Split(manifest, "\n") {
```

Comment:
wouldn't it be better to use something like the following instead of splitting the text along `\n`s?
```
scanner := bufio.NewScanner(strings.NewReader(manifest))
for scanner.Scan() {
   line := scanner.Text()
   ...
}
```

Reply:
Good idea - done.